### ISSUE TITLE
Analyzer/FE integration has some test failures that appear to be due to bad tests

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#34047]
It appears that a nontrivial fraction of failing test cases in the analyzer/FE integration are due to tests that are incorrect.  (In a random sampling of 60 of the ~500 failures currently outstanding, 4 appeared to be due to bad tests).  These tests need to be evaluated, and if they are indeed bad, fixed.

Suspected bad tests found so far:
- language_2/type_promotion_functions_test
  - This whole test seems to be bad (has an unconditional compile time error at line 100)
- language_2/override_inheritance_method_test/29
  - This test attempts to pair a base class method of type (int) -> dynamic with an override of type (String) -> dynamic, but does not expect an error
- language_2/string_supertype_checked_test
  - This test contains a class that implements Comparable<dynamic>, therefore the base class declaration of compareTo has type (dynamic) -> int, but the override has type (Comparable) -> int, which is an invalid override

I'll keep this list updated if I discover more bad tests.

**Comment by devoncarew** [comment]
@leafpetersen and @lrhn - it may be useful to triage the failing language_2 tests to try and identify bad tests - esp. for tests like the first on Paul's list. I believe the dart 2 test migration effort did leave some amount of known work behind.

**Comment by eernstg** [comment]
Right, `type_promotion_functions_test` is broken&mdash;it relies on the Dart 1 `<<` relation. Should be rewritten such that the stated `<<` relations become actual Dart 2 subtype relationships (e.g., `FuncDynToDyn << FuncAtoDyn` becomes `FuncNullToDyn <: FuncAtoDyn` where `typedef FuncNullToDyn(Null n);`).

Similarly, I agree that `override_inheritance_method_test/29` should be adjusted to `//# 29: compile-time error'.

In `string_supertype_checked_test`, the comment refers to 'supertypes of String' which seems to have very little to do with the source code that follows, but one way to make the code meaningful today would be to use `implements Comparable<A>` and `int compareTo(Comparable<A> other)`, which preserves the properties that the override uses variance and `promote` performs a cast.

> the dart 2 test migration effort did leave some amount of known work behind.

Indeed, except perhaps for 'known [by all]'. ;)

**Comment by kmillikin** [comment]
As Erik says, it is a known issue that some of the language_2 tests are bad tests.  There was a issue in the tracker that they needed to be triaged, from at least a year ago, but I can't find that issue right now (maybe it's closed?) and as far as I can remember it never really had an owner anyway.

The usual thing we've been doing is fixing them as we find them.  Beware that it's very painful to do that due to the status file updates required.

**Comment by stereotype441** [comment]
I've now finished triaging the remaining test.py failures for analyzer/CFE integration, so I believe the list above is reasonably complete.

**Comment by fishythefish** [comment]
I'm currently fixing tests that are failing on large numbers of bots, including several of the ones listed. Many of these are `language_2` tests that have a discrepancy between analyzer and CFE or which have been broken since the Dart 2 migration.
